### ISSUE TITLE
[NCO Bug] Fix compilation error with "-check all" for gfs_bufrsnd

### ISSUE THREAD

**Issue by KateFriedman-NOAA** [NOAA-EMC/global-workflow#1245]
Bugzilla: [#1208](http://www2.spa.ncep.noaa.gov/bugzilla/show_bug.cgi?id=1208)

NCO requested that `-check all` option be corrected to work with `gfs_bufrsnd` code. Details from bugzilla below:

```
We tested the gfs.v16.0.7 source code to compile with "-check all" option,  the gfs_bufrsnd failed during the compilation. 

Here is the error:

"meteorg.f(1108): error #5560: Subscript #1 of the array ZP2 has value 11 which is greater than the upper bound of 2
       print*,'finish computing zp2(11-12) ', zp2(11),zp2(12)
----------------------------------------------^
meteorg.f(1108): error #5560: Subscript #1 of the array ZP2 has value 12 which is greater than the upper bound of 2
       print*,'finish computing zp2(11-12) ', zp2(11),zp2(12)
------------------------------------------------------^
compilation aborted for meteorg.f (code 1)
make: *** [meteorg.o] Error 1
"

Please check your code with "-check all" option and fix the problem by next upgrade. 

Notes that the "-check all" is not used to compile in production running code. Just for code stability test.

Please change line 1108 from
 print*,'finish computing zp2(11-12) ', zp2(11),zp2(12)
to 
 print*,'finish computing zp2(1-2) ', zp2(1),zp2(2)
```

**Comment by HuiyaChuang-NOAA** [comment]
@ShelleyMelchior-NOAA I believe this should be assigned to Bo Cui.
I am not certain if I am permitted to assign github issue to a contractor so am including you in this case

**Comment by ShelleyMelchior-NOAA** [comment]
Ok, thanks for the notice, @HuiyaChuang-NOAA . I don't think I have permissions to identify an asignee. @KateFriedman-NOAA , I am Bo's contract branch leader. Can you assign this github issue to @BoCui-NOAA for me?

**Comment by KateFriedman-NOAA** [comment]
@ShelleyMelchior-NOAA Given your request, I have assigned this issue to @BoCui-NOAA . Thanks!

**Comment by BoCui-NOAA** [comment]
@KateFriedman-NOAA Got the assigned! Thanks! What's the schedule to give  you  the updated gfs_bufsnd?

**Comment by KateFriedman-NOAA** [comment]
@BoCui-NOAA Would the end of FY23Q2 (end of March) be reasonable to you?

**Comment by BoCui-NOAA** [comment]
> @BoCui-NOAA Would the end of FY23Q2 (end of March) be reasonable to you?

Sure, work for me.  I will test the "-check all" option. Do you know which file and line that NCO adds this option for code stability test? Thanks!

**Comment by HuiyaChuang-NOAA** [comment]
@BoCui-NOAA I know you're working on NAEFS upgrade to be implemented at the end of year so please feel free to speak up if this update to GFS Bufr needs to be postponed.

**Comment by BoCui-NOAA** [comment]
@HuiyaChuang-NOAA  I'll speak up and postpone the test if I can't meet the due date, thanks very much for your consideration.

**Comment by KateFriedman-NOAA** [comment]
> Do you know which file and line that NCO adds this option for code stability test?

@BoCui-NOAA Do you have access to the bugzillas? That would be a good place to ask NCO. If you don't, I can ask in there.

**Comment by BoCui-NOAA** [comment]
> 

@KateFriedman-NOAA I already accessed to the bugzillas and asked the question. Thanks!

**Comment by WalterKolczynski-NOAA** [comment]
Since the source code is now in [NOAA-EMC/gfs-utils](https://github.com/NOAA-EMC/gfs-utils), there should be a companion issue opened there for the work (although I think we want to move it to UFS-Utils).

**Comment by BoCui-NOAA** [comment]
@KateFriedman-NOAA @HuiyaChuang-NOAA I fix the compilation error in gfs_bufrsnd. One file was modified and  I saved it on Dogwood at:

/lfs/h2/emc/vpppg/noscrub/bo.cui/BUFR/bufr/sorc/gfs_bufr.fd/meteorg.f

The test is based on current gfs package gfs.v16.3.7. There is no need to modify file "makefile_module" as Jen mentioned that the "-check all" is not used to compile in production running code. Just for code stability test.

**Comment by KateFriedman-NOAA** [comment]
@BoCui-NOAA Excellent! We should get this into the GFSv17 system for sure. Would you mind opening an issue and submitting a PR into the https://github.com/NOAA-EMC/gfs-utils repo? We recently moved the gfs_bufr.fd code into that repo for GFSv17.

@arunchawla-NOAA @aerorahul Do we want to get this fix into the GFSv16 system as well? There will be an operational update for a GSI update in the coming weeks.

**Comment by BoCui-NOAA** [comment]
@KateFriedman-NOAA I did a quick comparison of gfs_bufr package between gfs.v16.3.7 and gfsv17. There are some differences between them. One example:

line 80 of https://github.com/BoCui-NOAA/gfs-utils/blob/develop/src/gfs_bufr.fd/calpreciptype.f
line  80 of https://github.com/NOAA-EMC/global-workflow/blob/dev/gfs.v16/sorc/gfs_bufr.fd/calpreciptype.f

Are the latest gfs_bufrsnd version cloned to the repo of GFSv17?

**Comment by KateFriedman-NOAA** [comment]
@BoCui-NOAA Those two `gfs_bufr.fd` folders you reference are the latest versions. The gfs-utils repo one is for GFSv17 and the global-workflow repo `dev/gfs.v16` branch one is the current GFSv16 operational version.

For GFSv17, when we clone the global-workflow `develop` branch we get the gfs-utils repo when we run `sorc/checkout.sh`.

Let me know if I didn't answer your question. Thanks!

**Comment by BoCui-NOAA** [comment]
@KateFriedman-NOAA Please compare  line 80  of the same file from the two folders. One is commented, which is the operational version and was implemented last June. Another one from gfs-utils repo was not commented.  So I am wondering if the gfs-utils repo hasn't gotten the latest version. Please let me know. Thanks!

**Comment by KateFriedman-NOAA** [comment]
@BoCui-NOAA Oh I see, yes I would assume the gfs-utils does not have the latest version that is in ops. Would you mind also committing that update into the gfs-utils copy of the code, along with the compilation error fix in both repos? Thanks!

**Comment by KateFriedman-NOAA** [comment]
@BoCui-NOAA Checking in on this.

**Comment by BoCui-NOAA** [comment]
@KateFriedman-NOAA I committed the changes to my git account.

https://github.com/BoCui-NOAA/global-workflow/commit/aa49fd1d7f85af6fb7242de1c241d0526b75ae60

Should I submit a PR to https://github.com/NOAA-EMC/global-workflow/tree/dev/gfs.v16 ?

**Comment by BoCui-NOAA** [comment]
@KateFriedman-NOAA For the GFS v17,  I compared  two gfs_bufr.fd folders between  gfs-utils repo and global-workflow repo dev/gfs.v16 branch. There are some other differences that cannot be traced back.  
https://github.com/NOAA-EMC/global-workflow/blob/develop/sorc/checkout.sh
Did you run this script to clone bufr package to gfs-utils repo? Thanks!

**Comment by KateFriedman-NOAA** [comment]
> Should I submit a PR to https://github.com/NOAA-EMC/global-workflow/tree/dev/gfs.v16 ?

Yes, please submit the bugzilla fix to that global-workflow branch for the GFSv16 package. Thanks!

Please also submit a PR to the gfs-utils `develop` branch to get gfs_bufr.fd updates into that repo for GFSv17. Yes, we clone that repo via global-workflow `checkout.sh`. Please make sure to refer to global-workflow issue #1245 in each PR. Thanks!

**Comment by BoCui-NOAA** [comment]
@KateFriedman-NOAA I submitted  a PR to GFS v16 package. But I missed to refer issue #1245. Do you know how to add it? Thanks!

**Comment by BoCui-NOAA** [comment]
@KateFriedman-NOAA I added issue #1245 in the PR to GFSv16. Please check if it it ok. Thanks!

**Comment by KateFriedman-NOAA** [comment]
@BoCui-NOAA Thanks for submitting PR #1580! I have approved and merged it into the `dev/gfs.v16` branch. This completes that part of resolving this issue for the GFSv16 package. The upcoming GFSv16.3.8 hand-off to NCO will include this bugzilla fix.

Just need it in the GFSv17 package now via the gfs-utils repo.

**Comment by BoCui-NOAA** [comment]
@KateFriedman-NOAA I updated gfs_bufr.fd and tocsbufr.fd using the gfs.v16 latest version to https://github.com/BoCui-NOAA/gfs-utils

There are lot of differences between gfs.v16 version and gfs-utils repo. Is this reasonable?

**Comment by KateFriedman-NOAA** [comment]
Pinging @aerorahul. The gfs-utils was cmake'd and had some other updates. @BoCui-NOAA can you list the diffs for gfs_bufr code that need consideration or that make it difficult to insert the bugzilla fix? Thanks!

**Comment by aerorahul** [comment]
@BoCui-NOAA 
We did not make any update to the Fortran source code (not intentionally anyway) when we extracted the utilities into the gfs-utils repository.
We updated the build system to CMake.

**Comment by BoCui-NOAA** [comment]
@KateFriedman-NOAA  @aerorahul  I found that gfs-utils is not using the latest version gfs_bufr.fd from gfs.v16. 

Here is a example,

line 80 of https://github.com/NOAA-EMC/global-workflow/blob/dev/gfs.v16/sorc/gfs_bufr.fd/calpreciptype.f
!     call gfuncphys

line 80 of https://github.com/NOAA-EMC/gfs-utils/blob/develop/src/gfs_bufr.fd/calpreciptype.f
      call gfuncphys

Would you please check which version you are use to extract gfs_bufr.fd and tocsbufr.fd? Thanks!

**Comment by KateFriedman-NOAA** [comment]
@BoCui-NOAA If gfs-utils doesn't have the latest gfs_bufr code then please include updates in a gfs-utils PR to get it using the latest version, as well as the bugzilla fix. Thanks!

**Comment by BoCui-NOAA** [comment]
> @BoCui-NOAA Thanks for submitting PR #1580! I have approved and merged it into the `dev/gfs.v16` branch. This completes that part of resolving this issue for the GFSv16 package. The upcoming GFSv16.3.8 hand-off to NCO will include this bugzilla fix.
> 
> Just need it in the GFSv17 package now via the gfs-utils repo.

Great, thanks @KateFriedman-NOAA

**Comment by BoCui-NOAA** [comment]
@aerorahul @KateFriedman-NOAA I created a clean and latest version of gfs_bufr.fd, which  include the bigzilla fix and Cmake is kept. The PR was submitted to gfs-utils.

**Comment by KateFriedman-NOAA** [comment]
Thanks @BoCui-NOAA !

**Comment by KateFriedman-NOAA** [comment]
New gfs-utils hash with bugzilla fix: [8965258](https://github.com/NOAA-EMC/gfs-utils/commit/896525808fc83b2d5690224c985c2c6f9ab6451b)
